[motherless] Add gallery support

Ti4eeT4e opened the pull request:
<!--
# Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

Fixes the pattern that the MotherlessIE uses for IDs. The site uses hexadecimal ids, so we can restrict the match here appropriately. Implements supports for galleries, which uses IDs prefixed with `G` or `GV`, which in the previous step we prevented the generic IE from matching.

Commit:
motherless: restrict valid url for MotherlessIE

IDs are hexadecimal and this reduces mismatches with the gallery IE.

Commit:
motherless: implement MotherlessGalleryIE

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
+        webpage = self._download_webpage(page_url, gallery_id)
+        title = self._search_regex(
+            r'<title>([\w\s]+\w)\s+', webpage, 'title', fatal=False)
```
self._html_extract_title

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+            webpage, 'page_count', default=1), 'page_count')
+        PAGE_SIZE = 60
+
+        def _get_page(idx):
+            if not page_count:
+                return
+            webpage = self._download_webpage(
+                page_url, gallery_id, query={'page': idx + 1},
+                note='Downloading page %d/%d' % (idx + 1, page_count)
+            )
+            for entry in self._extract_entries(webpage, url):
+                yield entry
```
imo, this should just be part of `_extract_entries`

Ti4eeT4e replied:
Which bits exactly? This is the pager function that is being passed into `InAdvancedPagedList`.

pukkandan replied:
The whole thing. You can use `functools.partial` as needed

Ti4eeT4e replied:
Do you have an example extractor I can look at for that?

pukkandan replied:
ping me once other points are addressed, and I'll help with this

Ti4eeT4e replied:
I think most other points are.

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+        PAGE_SIZE = 60
+
+        def _get_page(idx):
+            if not page_count:
+                return
```
how can this happen?

pukkandan replied:
ah, I now see how similar the codes for the 2 extractors are. Why not generalize it into a single base IE?

Ti4eeT4e replied:
I'll look into that.

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+                page_url, gallery_id, query={'page': idx + 1},
+                note='Downloading page %d/%d' % (idx + 1, page_count)
+            )
+            for entry in self._extract_entries(webpage, url):
+                yield entry
```
```suggestion
            yield from self._extract_entries(webpage, url)
```

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+        def _get_page(idx):
+            if not page_count:
+                return
+            webpage = self._download_webpage(
+                page_url, gallery_id, query={'page': idx + 1},
+                note='Downloading page %d/%d' % (idx + 1, page_count)
+            )
```
```suggestion
            webpage = self._download_webpage(
                page_url, gallery_id, query={'page': idx + 1},
                note='Downloading page %d/%d' % (idx + 1, page_count))
```

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+    def _real_extract(self, url):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
+        webpage = self._download_webpage(page_url, gallery_id)
```
This webpage is being downloaded twice?

Ti4eeT4e replied:
To populate the metadata and `page_count`

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+            for entry in self._extract_entries(webpage, url):
+                yield entry
+
+        playlist = InAdvancePagedList(_get_page, page_count, PAGE_SIZE)
+
+        return {
+            '_type': 'playlist',
+            'id': gallery_id,
+            'title': title,
+            'entries': playlist
+        }
```
```suggestion
        return self.playlist_result(
            InAdvancePagedList(_get_page, page_count, PAGE_SIZE), gallery_id, title)
```

Ti4eeT4e commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+
+    def _real_extract(self, url):
+        gallery_id = self._match_id(url)
+        page_url = compat_urlparse.urljoin(url, f'/GV{gallery_id}')
```
Because it would shadow the `url` var passed from the argument list, but that is probably not a problem after matching the id.

pukkandan replied:
I meant: why construct `page_url` and not directly use the `url`

Ti4eeT4e commented on yt_dlp/extractor/motherless.py:
```diff
@@ -246,3 +246,74 @@ def _get_page(idx):
+        page_count = self._int(self._search_regex(
+            r'(\d+)</a><a[^>]+rel="next">',
+            webpage, 'page_count', default=1), 'page_count')
+        PAGE_SIZE = 60
```
Done

pukkandan commented:
pls do not force-push. Having a messy commit history is fine

Commit:
motherless: Extract pagination into abstract base class

Commit:
motherless: update group and gallery url pattern

The abstraction does not yet handle these variations.

For galleries we need to paginate on /GV<ID>?page=n, but ideally we
could also just accept /G<ID> in addition to /GV<ID>.

Same goes for groups, where /g/<NAME> refers to the group and /gv/<NAME>
for the groups videos, where the pagination happens.

Commit:
motherless: drop ABC parent class

Ti4eeT4e commented:
Dropped ABC from the PaginatedIE class to fix the tests.

Commit:
motherless: add whitespaces around arithmetic operators

Ti4eeT4e commented:
Fixed flake8 errors.

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
+
+    def _extract_entries(self, webpage, base):
+        for mobj in re.finditer(r'href=".*(?P<href>\/[A-F0-9]+)"\s+title="(?P<title>[^"]+)', webpage):
+            video_url = compat_urlparse.urljoin(base, mobj.group("href"))
```
```suggestion
            video_url = compat_urlparse.urljoin(base, mobj.group('href'))
```

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
+                    video_title=mobj.group('title'))
+
+    def _real_extract(self, url):
+        _id = self._match_id(url)
```
Conventionally, we use `item_id`, but 🤷

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,88 @@ def _real_extract(self, url):
+    def _real_extract(self, url):
+        _id = self._match_id(url)
+        webpage = self._download_webpage(url, _id)
+        title = self._search_regex(r'^([\w\s]+)\s+', self._html_extract_title(webpage), 'title')
```
Why the search_regex?

Ti4eeT4e replied:
Title looks like so:

```
<title>F3C5049 | MOTHERLESS.COM ™</title>
```

I could also

```
title = "|".join(self._html.extract_title(webpage).split('|')[:-1]
```

Ti4eeT4e replied:
Actually also like so:

```html
<title>Movie Scenes - Videos - Hot and sexy scenes from &quot;regular&quot; movies... Beautiful actresses fully | MOTHERLESS.COM ™</title>
```

where `Movie Scenes` would be the title we're looking for, so the regex works.

pukkandan replied:
Why is "Movie Scenes" the title here? The whole thing except the suffix should be the title. So do `strip_or_none(remove_end(..., '| MOTHERLESS.COM '))`

Ti4eeT4e replied:
It is three things:

- Movie scenes (title)
- Video (meaning videos, as opposed to images)
- Hot and sexy scenes .... (description)

But that is handled inconsistently between groups and galleries.

Commit:
motherless: use single quotes

Ti4eeT4e commented:
Tests succeed, anything else?

pukkandan committed:
fix

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
-            entries.append(self.url_result(
-                video_url, ie=MotherlessIE.ie_key(), video_id=video_id,
-                video_title=title))
-        # Alternative fallback
-        if not entries:
-            entries = [
-                self.url_result(
-                    compat_urlparse.urljoin(base, '/' + entry_id),
-                    ie=MotherlessIE.ie_key(), video_id=entry_id)
-                for entry_id in orderedSet(re.findall(
-                    r'data-codename=["\']([A-Z0-9]+)', webpage))]
```
why was this removed

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
-        webpage = self._download_webpage(page_url, group_id)
-        title = self._search_regex(
-            r'<title>([\w\s]+\w)\s+-', webpage, 'title', fatal=False)
-        description = self._html_search_meta(
-            'description', webpage, fatal=False)
```
same

pukkandan commented on yt_dlp/extractor/motherless.py:
```diff
@@ -147,102 +146,81 @@ def _real_extract(self, url):
-        page_count = self._int(self._search_regex(
-            r'(\d+)</(?:a|span)><(?:a|span)[^>]+rel="next">',
-            webpage, 'page_count', default=0), 'page_count')
-        if not page_count:
-            message = self._search_regex(
-                r'class="error-page"[^>]*>\s*<p[^>]*>\s*(?P<error_msg>[^<]+)(?<=\S)\s*',
-                webpage, 'error_msg', default=None) or 'This group has no videos.'
-            self.report_warning(message, group_id)
```
this too

bashonly commented:
Closed by f2ff0f6f1914b82d4a51681a72cc0828115dcb4a
